### PR TITLE
AD-3 (Implement React and SignalR)

### PR DESCRIPTION
**Pull request by MinhLu12**
To get the application running:
1. Make sure web API is running on IIS
2. `npm start` in the client/ directory
3. Message me for the password for login (we can figure out the best way to transfer that, maybe just in Slack for now?)

Reference to the design document: https://docs.google.com/document/d/1SUOwBpVA55hXOANGyHlSXm_SZFIHivj3G8LcsxXD0Jk/edit

### CODE REVIEW

File: AdminDashboard/BusinessLogicOrchestrators/AccountOrchestrator/AccountOrchestrator.cs [hunk 1]
```diff
@@ -35,6 +40,8 @@ public async Task RegisterUserToAccount(Guid accountId, Guid userId)
                 throw new UserLimitExceededException();
 
             await Repository.AddUser(accountId, userId);
+
+            await UserHubContext.Clients.All.SendAsync("AddedUser");
```
Comment (MinhLu12) [inline AdminDashboard/BusinessLogicOrchestrators/AccountOrchestrator/AccountOrchestrator.cs:44]: This initiates a server side request to tell the client that another user has been registered

File: AdminDashboard/Models/Domain/Account.cs [hunk 2]
```diff
@@ -37,7 +37,9 @@ public EnderUserAccountRequest ToEndUser()
             {
                 Id = Id,
                 Users = Users,
-                CurrentPlan = CurrentPlan.GetPlanType()
+                CurrentPlan = CurrentPlan.GetPlanType(),
+                MaximumNumberOfUsersAllowed = CurrentPlan.MaxNumberOfUsersAllowed,
+                PricePerMonth = CurrentPlan.PricePerMonth
             };
```
Comment (MinhLu12) [inline AdminDashboard/Models/Domain/Account.cs:43]: Change contract for end users - we'll need some other fields from the API. This is so we don't need to do business logic in the client side.

File: client/src/Dashboard/Dashboard.jsx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+import { authenticationRepository } from '@/_services';
+import config from 'config';
+const signalR = require('@aspnet/signalr')
+
+class Dashboard extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            accountId: null,
```
Comment (MinhLu12) [inline client/src/Dashboard/Dashboard.jsx:14]: If the account object is used in more than one component, I'll make a model for it!

File: client/src/Dashboard/Dashboard.jsx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+    componentDidMount() {
+        const hub = this.configureUserHub();
+        this.start(hub);
+
+        this.createAndGetAccount();
+    }
+
+    createAndGetAccount() {
+        accountRepository.create()
```
Comment (MinhLu12) [inline client/src/Dashboard/Dashboard.jsx]: Every time we go to the Dashboard, we'll make an account and retrieve it for the user since there's no way in the UI to make an account. In the future I'd probably expose a button in the project to make an account first.

File: client/src/_services/account.service.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import config from 'config';
```
Comment (MinhLu12) [inline client/src/_services/account.service.js:1]: The following two files are the repository layers to the API

File: AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs [hunk 6]
```diff
@@ -68,7 +68,7 @@ private static byte[] GetBytesFrom(string savedPasswordHash)
 
         private bool UsernameIsIncorrect(string username)
         {
-            return Username != username;
+            return username != Username;
```
Comment (pierrebeaucamp) [inline AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs]: Why the change here?

File: client/.vscode/launch.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
```
Comment (pierrebeaucamp) [inline client/.vscode/launch.json]: I know that a few JSON parsers allow comments, but I'd prefer to remove them here - it's technically invalid JSON and those default comments don't add any useful information to the file

File: client/src/Dashboard/Dashboard.jsx [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+            .then(id => accountRepository.get(id))
+            .then(createdAccount => {
```
Comment (pierrebeaucamp) [inline client/src/Dashboard/Dashboard.jsx]: I don't like the pattern of "creating" account, then getting it.

* There is no "Creation" of any record which takes place. We simply want to retrieve an existing record (the account) using our credentials.
* Even if a `create` call is needed, there is no reason why it should be followed by a `get`. The first call can just return the resource it just created.

Anyway, I think the core problem is that you don't pass any state from the Login call to this page. The login call could just return the account ID which you then use here to retrieve the record. Another option would be to simply call the `get` endpoint without an ID and let the backend retrieve it based on your authentication.

File: client/src/Dashboard/Dashboard.jsx [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        this.setState({ currentPlan: 2, maximumNumberOfUsersAllowed: 1000, pricePerMonth: 1000});
+    }
+
+    renderHeader() {
+        if (this.state.currentPlan == 1) {
+            return (
+                <header>Startup Plan - ${this.state.pricePerMonth}/Month</header>
+            );
+        }
+        else if (this.state.currentPlan == 2) {
```
Comment (pierrebeaucamp) [inline client/src/Dashboard/Dashboard.jsx]: Small nitpick: The `else` here is redundant as the previous clause will return

File: AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs (hunk 6 above)
Comment (MinhLu12) [reply AdminDashboard/BusinessLogicOrchestrators/LoginOrchestrator/LoginOrchestrator.cs]: My bad, I tried something local and didn't revert it back 100% how it was, I'll change it!

File: client/.vscode/launch.json (hunk 7 above)
Comment (MinhLu12) [reply client/.vscode/launch.json]: Removed

File: client/src/Dashboard/Dashboard.jsx (hunk 9 above)
Comment (MinhLu12) [reply client/src/Dashboard/Dashboard.jsx]: Yeah I tried to be explicit that there's plans 1 and 2, but good point, in the end it's just redundant

File: client/src/Dashboard/Dashboard.jsx (hunk 8 above)
Comment (MinhLu12) [reply client/src/Dashboard/Dashboard.jsx]: Great point, making changes.

File: client/package.json [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    },
+    "dependencies": {
+        "@aspnet/signalr": "^1.1.4",
+        "formik": "^1.5.8",
+        "history": "^4.10.1",
+        "react": "^16.11.0",
+        "react-dom": "^16.11.0",
+        "react-router-dom": "^5.1.2",
+        "rxjs": "^6.5.3",
+        "yup": "^0.27.0"
```
Comment (alex-kovoy) [inline client/package.json]: I think some of it can be removed

File: client/src/Dashboard/Dashboard.jsx [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+                <header className="top-nav">
+                <h1>
+                    <i className="material-icons">supervised_user_circle</i>
+                    User Management Dashboard
+                </h1>
+                <button onClick={this.logout} className="button is-border">Logout</button>
+                </header>
+                
+                {this.renderUserLimitExceededMessage()}
+                {this.renderUpgradeSuccessMessage()}
```
Comment (pierrebeaucamp) [inline client/src/Dashboard/Dashboard.jsx:126]: This seems like it's always rendering the "upgrade success" message, even if the user is re-visiting the page at a later time

File: client/src/index.jsx [hunk 12]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import { render } from 'react-dom'
+
+import { LoginPage } from './LoginPage';
+import { Dashboard } from './Dashboard';
+import { Switch, Route, BrowserRouter as Router } from 'react-router-dom'
+
+const routing = (
+  <Router>
+    <div>
```
Comment (alex-kovoy) [inline client/src/index.jsx]: Do we need this div?

File: client/src/Dashboard/Dashboard.jsx [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+        }
+    }
+
+    upgradePlan() {
+        accountRepository.upgradePlan(this.state.accountId);
+        this.setState({ currentPlan: 2, maximumNumberOfUsersAllowed: 1000, pricePerMonth: 1000});
+    }
+
+    renderHeader() {
+        if (this.state.currentPlan == 1) {
```
Comment (pierrebeaucamp) [inline client/src/Dashboard/Dashboard.jsx]: Small nit: Identifying a plan just by the integer can be confusing and lead to bugs. Personally, I would prefer this to be a string which would carry more meaning with it.

As a side effect, it would also eliminate this entire method:
```jsx
<header>${this.state.currentPlan} Plan - ${this.state.pricePerMonth}/Month</header>
```

File: client/src/LoginPage/LoginPage.jsx [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+    handlePasswordChange(event) {
+        this.setState({ password: event.target.value });
+    }
+
+    handleSubmit(event) {
+        event.preventDefault();
+
+        authenticationRepository.login(this.state.username.split('@')[0], this.state.password)
+            .then(() => {
```
Comment (pierrebeaucamp) [inline client/src/LoginPage/LoginPage.jsx:31]: This is an improvement to your previous version, but I still don't get why we need to call `accountRepository.create()` here.

Would it be possible for `authenticationRepository.login` to return the ID of the account that is trying to log in?

Also it would be nice to have an error handler for `login` as entering wrong credentials is a common error.

File: client/src/LoginPage/LoginPage.jsx [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+                    this.props.history.push({
+                        pathname: '/Dashboard',
+                        state: { id: id }
+                    })
+                })
+            });
+      }
+
+    render() {
+        return (
```
Comment (alex-kovoy) [inline client/src/LoginPage/LoginPage.jsx:45]: Consider declaring the variables used in jsx up-front. This way it clearly indicates which state props are used inside jsx.

```jsx
const { username, password} = this.state;
```

File: client/src/LoginPage/LoginPage.jsx [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    handlePasswordChange(event) {
+        this.setState({ password: event.target.value });
+    }
+
+    handleSubmit(event) {
+        event.preventDefault();
+
+        authenticationRepository.login(this.state.username.split('@')[0], this.state.password)
+            .then(() => {
+                accountRepository.create()
```
Comment (alex-kovoy) [inline client/src/LoginPage/LoginPage.jsx:32]: No sure why this `accountRepository.create()` is called here. It could have returned the id on `login`

File: client/src/LoginPage/LoginPage.jsx [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    handlePasswordChange(event) {
+        this.setState({ password: event.target.value });
+    }
+
+    handleSubmit(event) {
+        event.preventDefault();
+
+        authenticationRepository.login(this.state.username.split('@')[0], this.state.password)
```
Comment (alex-kovoy) [inline client/src/LoginPage/LoginPage.jsx:30]: No server side error handling

File: client/src/LoginPage/LoginPage.jsx [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+            .then(() => {
+                accountRepository.create()
+                .then(id => {
+                    this.props.history.push({
+                        pathname: '/Dashboard',
+                        state: { id: id }
```
Comment (alex-kovoy) [inline client/src/LoginPage/LoginPage.jsx:36]: There is no need to push this ID to the history state. You could use URL parameter for that.

File: client/src/LoginPage/LoginPage.jsx (hunk 14 above)
Comment (MinhLu12) [reply client/src/LoginPage/LoginPage.jsx:31]: Right now, there's no exposure on the UI to create the account resource, so the client has to make it somewhere. 

Are you suggesting we make that resource when the user is logging in the `Login` API call? That right now returns the token. I should add Account ID in the response too?

File: client/src/LoginPage/LoginPage.jsx [hunk 19]
```diff
@@ -0,0 +1,65 @@
+import React from 'react';
+import '../index.css';
+import { authenticationRepository, accountRepository } from '@/_services';
+
+class LoginPage extends React.Component {
```
Comment (alex-kovoy) [inline client/src/LoginPage/LoginPage.jsx:5]: Missing propTypes which is useful because you can see what props this component expects (like `history`)

File: client/src/LoginPage/LoginPage.jsx (hunk 14 above)
Comment (pierrebeaucamp) [reply client/src/LoginPage/LoginPage.jsx:31]: I think what confuses me is that we have create a resource at all. The account should already exist on the backend and we simply retrieve it here.

You seem to depend on three separate calls to do that:
1. Login and get an authentication token
2. "Create" a resource to get the ID of the account
3. Retrieve the account using the ID from step 2.

I think we should be able to remove at least one of those steps. I.e. the `login` API call could return the ID of our account. Another possibility would be to use the `get` API call without any ID and let the backend return the currently logged in account.

Or maybe I'm just understanding something incorrectly. Why exactly do we need to "create" / "make" that account resource first?

File: client/src/Dashboard/Dashboard.jsx (hunk 13 above)
Comment (MinhLu12) [reply client/src/Dashboard/Dashboard.jsx]: Created an enum for them so it makes more sense!

File: client/src/Dashboard/Dashboard.jsx [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+            hub: null
+        };
+
+        this.logout = this.logout.bind(this);
+        this.upgradePlan = this.upgradePlan.bind(this);
+    }
+
+    componentDidMount() {
+        const hub = this.configureUserHub();
+        this.start(hub);
```
Comment (alex-kovoy) [inline client/src/Dashboard/Dashboard.jsx:29]: Looks like server errors are ignored

File: client/src/Dashboard/Dashboard.jsx [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+
+        this.getAccount();
+    }
+
+    getAccount() {
+        accountRepository.get(this.props.location.state.id)
+            .then(createdAccount => {
```
Comment (alex-kovoy) [inline client/src/Dashboard/Dashboard.jsx]: `createdAccount` indicates that there are other types of accounts, should it be just `account` ?

File: client/src/Dashboard/Dashboard.jsx [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            accountId: null,
+            currentPlan: null,
+            numberOfUsers: null,
+            maximumNumberOfUsersAllowed: null,
+            pricePerMonth: null,
+            hub: null
```
Comment (alex-kovoy) [inline client/src/Dashboard/Dashboard.jsx]: why `hub` is part of the state, is it supposed to change?

File: client/src/_services/account.service.js [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+        .then(response => response.json())
+}
+
+function upgradePlan(id) {
+    const requestOptions = {
+        method: 'PUT',
+        withCredentials: true,
+        headers: { 'Content-Type': 'application/json',
+                   'Accept': 'application/json',
+                   'Authorization': `Bearer ${sessionStorage.getItem("Token")}` },
```
Comment (alex-kovoy) [inline client/src/_services/account.service.js]: a lot of code duplication around handling security headers and API calls in general.

File: client/src/_services/account.service.js [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+function get(id) {
+    const requestOptions = {
+        method: 'GET',
+        withCredentials: true,
+        headers: { 'Content-Type': 'application/json',
+                   'Accept': 'application/json',
+                   'Authorization': `Bearer ${sessionStorage.getItem("Token")}` }
+    };
+
+    return fetch(`${config.apiUrl}/api/account/${id}`, requestOptions)
```
Comment (alex-kovoy) [inline client/src/_services/account.service.js]: Missing proper error handling (internal server errors will be ignored) I suggest to compare `fetch` and `xhr` APIs to find out the differences.

File: client/src/LoginPage/LoginPage.jsx (hunk 14 above)
Comment (MinhLu12) [reply client/src/LoginPage/LoginPage.jsx:31]: I think I understand the issue: I'm assuming there can be multiple account resources, each one associated with a different admin, so that's why we have to make an account every time.

But it's true since we only really ever have one admin, our *GET* Account could be called without a guid query parameter, and get the default Account in the backend. But, that means we have to manually make that account through the API before we start the app. Would that be okay with you?

EDIT: I can have logic to make a default account in the `login` API call, and return both the token and account Id to the client, so the client doesn't ever have to call `createAccount`

File: client/src/Dashboard/Dashboard.jsx [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+                
+                {this.renderUserLimitExceededMessage()}
+                {this.renderUpgradeSuccessMessage()}
+
+                <div className="plan">
+                    {this.renderHeader()}
+
+                    <div className="plan-content">
+                        <div className="progress-bar">
+                        <div style={{width: (this.state.numberOfUsers/this.state.maximumNumberOfUsersAllowed) * 100 + '%'}} className="progress-bar-usage"></div>
```
Comment (alex-kovoy) [inline client/src/Dashboard/Dashboard.jsx]: I would not inline any logic in jsx. Instead I would declare the progress variable at the beginning of `render()`

```jsx
const { numberOfUsers, maximumNumberOfUsersAllowed } = this.state;
const progres = `${numberOfUsers/maximumNumberOfUsersAllowed * 100}%`;
```

Also the naming it too verbose, this looks shorter:

```jsx
const { userCount, userLimit } = this.state;
```

File: client/src/Dashboard/Dashboard.jsx (hunk 11 above)
Comment (MinhLu12) [reply client/src/Dashboard/Dashboard.jsx:126]: Yes! I'll introduce logic to dismiss it if the user clicks the banner so it won't show after.

**Review by alex-kovoy — APPROVED** [review]

File: client/src/index.jsx (hunk 12 above)
Comment (MinhLu12) [reply client/src/index.jsx]: Probably not...Deleted!

File: client/src/LoginPage/LoginPage.jsx (hunk 16 above)
Comment (MinhLu12) [reply client/src/LoginPage/LoginPage.jsx:32]: Pierre gave me the idea to make one account on startup, and simply GET it in the client, so then there'd be no need for a POST!

File: client/src/LoginPage/LoginPage.jsx (hunk 14 above)
Comment (pierrebeaucamp) [reply client/src/LoginPage/LoginPage.jsx:31]: Can we create that account when the backend starts up? So we don't need an extra API call for that.

If that takes too long, we can also keep it as-is for now, I was just wondering about the reason for this call

File: client/src/LoginPage/LoginPage.jsx (hunk 18 above)
Comment (pierrebeaucamp) [reply client/src/LoginPage/LoginPage.jsx:36]: Not sure if I would put in the URL - it would be weird to have an account ID in there, wouldn't it? I think the history state is fine for this example.

File: client/src/LoginPage/LoginPage.jsx (hunk 14 above)
Comment (MinhLu12) [reply client/src/LoginPage/LoginPage.jsx:31]: That's a good idea, I can definitely make a default account on start up, and then just GET it in the client! Making changes.

File: client/src/LoginPage/LoginPage.jsx (hunk 18 above)
Comment (alex-kovoy) [reply client/src/LoginPage/LoginPage.jsx:36]: History API is not a good place for storing this info because you wont be able to open another tab (as that state parameter will be empty). So there are 2 options:
1. return this ID with auth response and store it in the session/local storage.  
2. for simplicity reasons it could be placed in URL.

Comment (pierrebeaucamp) [reply client/src/LoginPage/LoginPage.jsx:36]: yeah, that's true. In my mind that would be fine for the scope of this challenge, but it comes with tradeoffs.

File: client/src/Dashboard/Dashboard.jsx (hunk 22 above)
Comment (MinhLu12) [reply client/src/Dashboard/Dashboard.jsx]: Good point, removed it from state.

File: client/src/Dashboard/Dashboard.jsx (hunk 25 above)
Comment (MinhLu12) [reply client/src/Dashboard/Dashboard.jsx]: True, they're too wordy. Changing.

File: client/src/LoginPage/LoginPage.jsx (hunk 14 above)
Comment (MinhLu12) [reply client/src/LoginPage/LoginPage.jsx:31]: I've spent a few hours on this and it looks like it's too much scope unfortunately. I got an account on the server upon start up, but since there's no concept of account ID anymore (aka multiple accounts), all the other endpoints through all layers have to change. That, and integration tests and client have to change as well.

I'll keep it for now if that's okay!

Comment (pierrebeaucamp) [reply client/src/LoginPage/LoginPage.jsx:31]: Of course. Don't worry about it too much, this application doesn't need to be perfect - we just want to see what tradeoffs were made and the thinking behind the architecture, etc. Your input and comments are very valuable.

File: client/src/_services/account.service.js [hunk 26]
```diff
@@ -7,21 +7,24 @@ export const accountRepository = {
 };
 
 function create() {
-    const requestOptions = getOptions('POST', JSON.stringify({ "plan": 1 }));
+    const requestOptions = getOptions('POST', JSON.stringify({ "plan": 2 }));
 
     return fetch(`${config.apiUrl}/api/account`, requestOptions)
         .then(response => response.json())
+        .catch(err => alert(err));
```
Comment (pierrebeaucamp) [inline client/src/_services/account.service.js]: I think a `console.log` would be better / less disruptive.

**Review by pierrebeaucamp — APPROVED** [review]
